internal/server/storage/storage.go

```diff
@@ -0,0 +1,191 @@
+package storage
+
+import (
+	"context"
+	"encoding/json"
+	"io/fs"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/npavlov/go-metrics-service/internal/domain"
+	"github.com/npavlov/go-metrics-service/internal/logger"
+	"github.com/npavlov/go-metrics-service/internal/model"
+	"github.com/npavlov/go-metrics-service/internal/server/config"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog"
+)
+
+const (
+	errNoValue = "no value provided"
+)
+
+type Number interface {
+	int64 | float64
+}
+
+type Repository interface {
+	Get(name domain.MetricName) (*model.Metric, bool)
+	Create(metric *model.Metric) error
+	GetAll() map[domain.MetricName]model.Metric
+	Update(metric *model.Metric) error
+	WithBackup(ctx context.Context, cfg *config.Config) *MemStorage
+}
+
+type MemStorage struct {
+	mu      *sync.RWMutex
+	metrics map[domain.MetricName]model.Metric
+	cfg     *config.Config
+	l       *zerolog.Logger
+}
+
+// NewMemStorage - constructor for MemStorage.
+func NewMemStorage() *MemStorage {
+	ms := &MemStorage{
+		metrics: make(map[domain.MetricName]model.Metric),
+		mu:      &sync.RWMutex{},
+		l:       logger.NewLogger().Get(),
+	}
+
+	return ms
+}
+
+func (ms *MemStorage) WithBackup(ctx context.Context, cfg *config.Config) *MemStorage {
+	ms.cfg = cfg
+
+	err := ms.restore()
+	if err != nil {
+		// Continue on error
+		ms.l.Error().Err(err).Msg("failed to restore metrics")
+	}
+
+	ms.l.Info().Msg("Metrics restored successfully")
+
+	ms.startBackup(ctx)
+
+	return ms
+}
+
+func (ms *MemStorage) startBackup(ctx context.Context) {
```

Comment:
[CHANGE REQUEST] We should not do it here. It would be better if you use `Memento`.
https://refactoring.guru/design-patterns/memento

Reply:
Refactored using this pattern

Reply:
[CHANGE REQUEST] Please follow the Open–closed principle